ls1intum/Artemis
# `Development`: Add certificates for e2e tests and improve e2e test flakiness detection

### Checklist
#### General


- [x] I tested **all** changes and their related features on a duplicated and modified E2E Bamboo Plan

### Motivation and Context
**E2E certificates**
This PR introduces a nginx reverse proxy with pre-configured certificates also configured in cypress, which should make the cypress workaround from the following PR not necessary anymore:

- https://github.com/ls1intum/Artemis/pull/5723

**E2E flake detection:**
The selected build for comparison, the last successful build, could previously end up being one later than the analyzed build.


### Description

**E2E certificates**
We started adopting a docker compose based nginx service based on the Ansible scripts from our current production nginx.
The certificates for this reverse proxy are configured in nginx and cypress and were created with the tool [mkcert](https://github.com/FiloSottile/mkcert).
These certificates are valid until 2025, but we also added a guide and helpers to generate them again.

This PR doesn't yet remove the workaround itself, as it's still needed for some local setups.

**E2E flake detection:**
This is a small change that adds the bamboo scripts into version control and forwards the current build number to the flake checker. This is to make sure the correct reference is used.

### Steps for Testing


... or at least look how we tested the CI pipeline for this branch.

We cloned the E2E test bamboo plans to our [own bamboo E2E test project/plan for testing](https://bamboo.ase.in.tum.de/browse/AMPU-AETM0-56) and disabled the hooks and scripts which interact with GitHub. Respectively, we did the same for [the E2E flake detection bamboo project/plan](https://bamboo.ase.in.tum.de/browse/AMPU-AETFDM0).

Then you can manually select this branch and run the modified CI pipeline (see Migration Path for the modified parts).

### Migration Path

Change the following tasks in the **[Artemis E2E Tests (Cypress in Docker-default)](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETG)** Bamboo Plan:
- Script `Deploy Artemis and run Cypress tests`: change to `Script location: File` and set it to `.bamboo/E2E-tests/execute.sh`
- Script `Make sure all docker containers are cleaned up`: change to `Script location: File` and set it to `.bamboo/E2E-tests/cleanup.sh`

Change the following tasks in the **[Artemis E2E (Cypress - Docker- Flaky)](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AECF)** Bamboo Plan:
- Script `Deploy Artemis and run Cypress tests`: change to `Script location: File` and set it to `.bamboo/E2E-tests-with-flake-detection/execute.sh`
- Script `Make sure all docker containers are cleaned up`: change to `Script location: File` and set it to `.bamboo/E2E-tests-with-flake-detection/cleanup.sh`

### Review Progress






#### Code Review
#### Manual Tests
